### DISCUSSION TITLE
How to create varibles for use in kdl config file

### DISCUSSION THREAD

**Discussion by chrishoage** [YaLTeR/niri#674] [Q&A]
I like to create templates that hold color information that I include into config files to be used in place of hard coded colors. 

This lets me more easily change or tweak themes with out having to use a dotfile manager that templates letting me more easily use stow to mange my dotfiles. 

I have looked though the wiki, open issues, and discussions, but I can't see how to do something like this, or even a way to support "includes" at all. 

Are either of these things supported and I've simply missed it, or have these features not been implemented? If not are there any future plans to implement these?

**Comment by YaLTeR** [comment]
Variables and includes are not as straightforward as they may be in other configs due to the declarative nature of niri's. I'm kind of reluctant with variables because even in a reasonably limited form, they would require replacing several automatically-derived parsers with manual code (to resolve the variables). Or if we decide that variables should work from anywhere in the config (which is generally how the rest of the config works), then config loading should become two-stage, first parsing, then resolving the variables. Sounds fairly annoying for the underlying types.

Includes, at least in a limited form, will happen at some point. But once again they would need non-trivial code changes.

**Reply by chrishoage** [reply]
What about reading and replacing environment variables instead of embedded variables? This would eliminate the issue of trying to deal with variables defined in the config itself. Could be something like `$env.color8` which replaces with whatever is in the environment that niri was launched in? 

I understand that the suggestion doesn't solve all of the problems, namely still requiring two stages in order to load the config.  I can definitely understand the complexity that variables bring. 

I wonder if it may be reasonable to severely limit where these "variables" may be used. 

e.g. have something like 

```
colors {
  border "#fff"
  active "#ccc"
}

window-rule {
  border {
    inactive-color $colors.border
    active-color $colors.active
  }
}
```

This could be included and these can only be referenced in highly specific areas, like colors, yet still be included into the config? 

(I don't love this that much to be honest)

**Reply by YaLTeR** [reply]
The two-stage loading sounds fairly annoying unfortunately because ideally the unresolved color type should be separate from the resolved color type (so that the rest of the code doesn't suddenly have to check if the color was resolved on every access), and it's deep inside several layers of config structs.

**Reply by chrishoage** [reply]
I keep coming back to this desire. 

I'm wondering if this approach would work: 

Niri would continue to not support includes or variables to keep the configuration language simple. 

The IPC would allow for changing color configuration at runtime. 

Users like myself who wish to co-locate theme configuration into a single spot while not having to remember to update a template instead of the actual dotfile could configure Niri to run a script on startup and dynamically set colors through the IPC. 

I took a look through the current IPC crate and it doesn't appear that setting configuration in this way is currently supported aside from being able to configure outputs temporarily. 

My proposal is allowing dynamic configuration of colors that could be set through the configuration file. 

To be clear just like output in IPC all changes would be forgotten and would be runtime only. 

Is this something that could be considered in the far future (since this would be entirely new functionality in the IPC) as a solution to this class of issue?

**Reply by bbb651** [reply]
I think it's a bit out of scope for niri, there are generic tools that allow for full scriptability like [ConfgenFS](https://git.mzte.de/LordMZTE/confgen#confgenfs) (see [blog](https://mzte.de/a/0003-confgenfs-all-config-files-are-now-scriptable/), this currently doesn't work with niri because [libfuse doesn't support fsnotify](https://git.mzte.de/LordMZTE/confgen/issues/8) so it needs some kind of `niri msg action reload-config` action to be added to manually trigger a reload) or [yolk](https://github.com/elkowar/yolk).

**Reply by YaLTeR** [reply]
> currently doesn't work with niri because libfuse doesn't support fsnotify

Our hot reload doesn't use notify at the moment

**Reply by chrishoage** [reply]
@bbb651 A key part of my post that I think was missed is I explicitly do not want a preprocessor for my configuration. An important part of my workflow is being able to edit the actual rc files that the software reads. With any sort of templating I am no longer able to do that. 

Nearly all of the other software to support window managers do offer some way to achieve this, but not Niri 

I am after some way to be able to do runtime configuration of Niri similar to the rest of its WM peers. 

I thought using the IPC to do runtime conflagration of Niri would be a happy compromise (and could even allow bspwm users like myself to configure niri from a shell script)

Anyway - just a thought. This is the one thing that is keeping me from fully switching to Niri from bspwm. Maybe one day!

**Reply by bbb651** [reply]
> @bbb651 A key part of my post that I think was missed is I explicitly do not want a preprocessor for my configuration. An important part of my workflow is being able to edit the actual rc files that the software reads. With any sort of templating I am no longer able to do that.

Yolk specifically has the property that files are always valid and directly editable, it instead embeds the code within comments. And in general a preprocessor doesn't inherently prohibit hot reloading, you just need an additional daemon to watch and recompile it.

The main reason I think variables are a bad idea is because it simultaneously does too much yet not enough: You very often want these variables to be synced with other programs (like colors with layer-shell clients), so now you either have to duplicate them and have multiple sources of truths that are cumbersome to edit and can get out of sync, or add a way to export variables and now niri suddenly becomes a state manager (there's a utility I forgot the name of that's explicitly manages state of variables like this, I generally believe in pragmatism over unix-ness and here it has the intimidate advantage of not tying your state to niri).

And you very often want some level of derived values - colors is a good example of this, e.g. making a gradient from a color, and it quickly explodes in complexity with CSS-like functions.

> I thought using the IPC to do runtime conflagration of Niri would be a happy compromise (and could even allow bspwm users like myself to configure niri from a shell script)

Imo this is much better, this already exists for outputs with `niri msg output`, this still introduces state within niri and but this is the minimal amount needed for allowing it to be controlled via a different tool/script. Although personally I still think it's not worth it, this introduces a lot of complexity, and what should happen when you edit the config while a value is overridden through IPC? Both ignoring it until next restart and overriding it can be confusing, should it display a warning?

**Reply by chrishoage** [reply]
> Yolk specifically has the property that files are always valid and directly editable, it instead embeds the code within comments. And in general a preprocessor doesn't inherently prohibit hot reloading, you just need an additional daemon to watch and recompile it.

Thanks! Sorry I didn't initially understand that was how it operates. That's an interesting solution but is complexity that I would prefer to avoid if possible. I will take a closer look at this tool to see if it fits for me however. 

> Although personally I still think it's not worth it, this introduces a lot of complexity, and what should happen when you edit the config while a value is overridden through IPC? Both ignoring it until next restart and overriding it can be confusing, should it display a warning?

Naively as a user I would expect "last change wins". 

For me personally I would only have a minimal config file (just enough to make Niri start) and then delegate configuration to the IPC like I do with `bspc`. 

I'm not sure it would be confusing if IPC changes were never persisted or if they were overridden by config file changes. I feel that most users who would read the documentation to learn about the IPC configuration options would read the note about how changes are runtime only and config reloads wipe out IPC changes. Those who don't will be confused by everything because they didn't RTFM :sweat_smile: 

In any case I can recognize the complexity concerns. I would however submit that offering IPC control for most/all configuration options really opens up Niri to be _highly_ scriptable/automatable which _could_ perhaps justify the complexity 

Thanks for humoring me in this discussion! :slightly_smiling_face:

**Comment by jjramsey** [comment]
I'd recommend that you use a pre-processor like m4. You can treat your "`.kdl.m4`" as the master that you edit, and then run m4 on it to generate the config file that Niri reads.

**Comment by chrishoage** [comment]
For anyone finding this thread in the future - includes were merged recently :) 

https://github.com/YaLTeR/niri/pull/2482
